File: 3434-find-the-number-of-distinct-colors-among-the-balls/3434-find-the-number-of-distinct-colors-among-the-balls.cpp
class Solution {
public:
    vector<int> queryResults(int limit, vector<vector<int>>& queries) {
        map<int, int> nums, colors;
        vector<int> res;

        for (auto q : queries) {
            int u = q[0], v = q[1];
            if (nums.count(u)) {
                colors[nums[u]]--;
                if (colors[nums[u]] == 0) {
                    colors.erase(nums[u]);
                }
            }
            nums[u] = v;
            colors[v]++;
            res.push_back(colors.size());
        }

        return res;
    }
};

File: 3434-find-the-number-of-distinct-colors-among-the-balls/3434-find-the-number-of-distinct-colors-among-the-balls.py
class Solution:
    def queryResults(self, limit: int, queries: List[List[int]]) -> List[int]:

        st = set()
        cnt = Counter()
        color = Counter()
        ans = [0] * len(queries)
        idx = 0
        for b, c in queries:
            if b in cnt:
                color[cnt[b]] -= 1
                if not color[cnt[b]]: st.remove(cnt[b])
                
            cnt[b] = c
            color[c] += 1
            st.add(c)
            ans[idx] = len(st)
            idx += 1
        return ans